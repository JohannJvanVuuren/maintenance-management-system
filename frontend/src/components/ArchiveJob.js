/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';

export const ArchiveJob = () => {

}