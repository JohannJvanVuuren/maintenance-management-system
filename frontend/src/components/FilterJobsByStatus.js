/* Import of axios client to make API calls to the backend of the application */
import axios from "axios";

/* Import of the React hooks used in this component */
import {useEffect, useState} from "react";

/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';
/* Import of React-Bootstrap styled component */
import Table from "react-bootstrap/Table";

export const FilterJobsByStatus = () => {
    /* Declaration and initialisation of the state variable needed in this component */
    const [jobs, setJobs] = useState([]);
    const [status, setStatus] = useState('');

    /* A useEffect hook to run the axios call to the backend to retrieve the required information from
    * the database */

    useEffect(() => {
        /* Setup of configuration of the axios call */
        const url = `http://localhost:8000/filterJobsByStatus/${status}`;

        /* axios call to the .../filterJobsByStatus endpoint to get the information that needs to be displayed
        * in this component */
        axios.get(url)
            .then(response => {
                setJobs(response.data);
                console.log(response);
                return response.data;
            })
            .catch(error => {
                console.log('Error', error.message);
            })
    }, [status]);

    return (
        <div className={'filter-by-status-container'}>
            <h1 className={'section-titles'}>Filter Jobs by status</h1>
            {/* Allowing the user to select a desired status and then setting the Status state to that choice */}
            <select value={status} onChange={(event) => {
                event.preventDefault();
                setStatus(event.target.value);
            }}>
                <option className={'items'} disabled={true}>--Please select a filter--</option>
                <option className={'items'} value={'Submitted'}>Submitted</option>
                <option className={'items'} value={'In progress'}>In progress</option>
                <option className={'items'} value={'Completed'}>Completed</option>
            </select>
            {/* Rendering of a React-bootstrap table suited for this purpose */}
            <Table
                striped bordered hover variant="dark"
                responsive
                className={'mt-5 table'}
            >
                <thead>
                <tr className={'table-header'}>
                    <th className={'h3'}>Unique Ref. No.</th>
                    <th className={'h3'}>Date</th>
                    <th className={'h3'}>Description</th>
                    <th className={'h3'}>Location</th>
                    <th className={'h3'}>Priority</th>
                    <th className={'h3'}>Status</th>
                </tr>
                </thead>
                <tbody>
                {/* The data is received back as an array of objects. So the Array.map method is used to map loop
                         through the array and display the data of the individual job objects */}
                {jobs.map((job, index) => {
                    /* Only non-archived documents are displayed */
                    if (job.archive === false) {
                    return (
                        <tr key={index} className={'table-row'}>
                            <td className={'h5'}>{job._id}</td>
                            <td className={'h5'}>{(job.date).slice(0, 10)}</td>
                            <td className={'h5'}>{job.description}</td>
                            <td className={'h5'}>{job.location}</td>
                            <td className={'h5'}>{job.priority}</td>
                            <td className={'h5'}>{job.status}</td>
                        </tr>
                    )} else {
                        return (
                        <tr>
                            <td colSpan={7}>
                            </td>
                        </tr>
                        )
                    }
                })}
                </tbody>
            </Table>
        </div>
    )
}