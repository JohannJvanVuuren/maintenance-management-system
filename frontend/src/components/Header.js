/* Import of the Link component from React Router to render functionality to the nav buttons */
import { Link } from 'react-router-dom';

/* Import of main stylesheet generated by the SCSS files and preprocessor */
import '../scss/main.css';

/* Definition of the Header components that will appear on all renders */
export const Header = () => {
    return (
        <div className={'header-container'}>
            <Link
                to={'/'}
                className={'header-title'}
            >
                <h1>Maintenance Management System (MMS)</h1>
            </Link>
            <div className={'header-navigation'}>
                <Link
                    to={'/submitJob'}
                >
                    <button className={'btn btn-nav'}>
                        Submit Job
                    </button>
                </Link>
                <Link
                    to={'/listJobs'}
                >
                    <button className={'btn btn-nav'}>
                        List Jobs By Date
                    </button>
                </Link>
                <Link
                    to={'/filterJobsByStatus'}
                >
                    <button className={'btn btn-nav'}>
                        Filter By Status
                    </button>
                </Link>
                <Link
                    to={'/updateInfoOfJob'}
                >
                    <button className={'btn btn-nav'}>
                        Update Job
                    </button>
                </Link>
                <Link
                    to={'/updateStatusMultipleJobs'}
                >
                    <button className={'btn btn-nav'}>
                        Update Status (Multiple)
                    </button>
                </Link>
                <Link
                    to={'/archiveJob'}
                >
                    <button className={'btn btn-nav'}>
                        Archive Job
                    </button>
                </Link>

            </div>
        </div>
    )
}