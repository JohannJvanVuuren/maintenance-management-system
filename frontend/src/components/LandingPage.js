/* Import of main stylesheet generated by the SCSS files and preprocessor */
import '../scss/main.css';

export const LandingPage = () => {
    return (
        <div className={'landing-container'}>
            <h1>Welcome</h1>
            <p>This system assists with the management of maintenance jobs and their tracking.</p>
            <p>Navigation through the system takes place though the navigation buttons at the top of the screen.</p>
            <p>To return to the home page click on the title at the top of the page.</p>
        </div>
    )
}