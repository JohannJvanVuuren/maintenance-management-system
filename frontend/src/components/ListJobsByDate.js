/* Import of the axios client to handle requests to the backend */
import axios from 'axios';

/* Import of React hooks used in this component */
import {useEffect, useState} from "react";

/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';
/* Import of React-Bootstrap styled components */
import Table from 'react-bootstrap/Table';

/* Definition of the ListJobsByDate component. The purpose of this component is to simply display
* all jobs in the database */
export const ListJobsByDate = () => {

    /* Declaration and initialisation of the state variable needed in this component */
    const [jobs, setJobs] = useState([]);

    /* A useEffect hook to run the axios call to the backend to retrieve the required information from
    * the database */
    useEffect(() => {

        /* Setup of configuration of the axios call */
        const url = 'http://localhost:8000/getJobListByDate';

        /* axios call to the .../getJobListByDate endpoint to get the information that needs to be displayed
        * in this component */
        axios.get(url)
            .then(response => {
                setJobs(response.data);
            })
            .catch(error => {
                console.log(error);
            })
    }, []); //No dependencies, so the hook will only run once when the page is rendered

    return (
        <div>
            <h1 className={'section-titles'}>List jobs by date</h1>
            {/* Rendering of a React-bootstrap table suited for this purpose */}
            <Table striped bordered hover variant="dark"
                   responsive
                   className={'mt-5 table'}
            >
                <thead>
                <tr className={'table-header'}>
                    <th className={'h3'}>Unique Ref. No.</th>
                    <th className={'h3'}>Date</th>
                    <th className={'h3'}>Description</th>
                    <th className={'h3'}>Location</th>
                    <th className={'h3'}>Priority</th>
                    <th className={'h3'}>Status</th>
                </tr>
                </thead>
                <tbody>
                {/* The data is received back as an array of objects. So the Array.map method is used to map loop
                 through the array and display the data of the individual job objects */}
                {jobs.map((job, index) => {
                    /* Only non-archived items will be listed */
                    if (job.archive === false) {
                        return (
                            <tr key={index} className={'table-row'}>
                                <td className={'h5'}>{job._id}</td>
                                <td className={'h5'}>{(job.date).slice(0, 10)}</td>
                                <td className={'h5'}>{job.description}</td>
                                <td className={'h5'}>{job.location}</td>
                                <td className={'h5'}>{job.priority}</td>
                                <td className={'h5'}>{job.status}</td>
                            </tr>
                        )
                    } else {
                        return (
                        <tr>
                            <td colSpan={7}>
                            </td>
                        </tr>
                        )
                    }
                })}
                </tbody>
            </Table>
        </div>
    )
}