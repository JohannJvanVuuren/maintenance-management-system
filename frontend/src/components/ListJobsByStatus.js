/* Import of axios client to make API calls to the backend of the application */
import axios from "axios";

/* Import of the React hooks used in this component */
import {useEffect, useState} from "react";

/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';
/* Import of React-Bootstrap styled component */
import Table from "react-bootstrap/Table";


export const ListJobsByStatus = () => {
    /* Declaration and initialisation of the state variable needed in this component */
    const [jobs, setJobs] = useState([]);

    /* A useEffect hook to run the axios call to the backend to retrieve the required information from
    * the database */
    useEffect(() => {

        /* Setup of configuration of the axios call */
        const url = 'http://localhost:8000/getJobListByStatus';

        /* axios call to the .../getJobListByStatus endpoint to get the information that needs to be displayed
        * in this component */
        axios.get(url)
            .then(response => {
                setJobs(response.data);
            })
            .catch(error => {
                console.log('Error', error.message);
            })
    }, []); //No dependencies, so the hook will only run once when the page is rendered

    return (
        <div>
            <h1 className={'section-titles'}>List Jobs by status</h1>
            {/* Rendering of a React-bootstrap table suited for this purpose */}
            <Table
                striped bordered hover variant="dark"
                responsive
                className={'mt-5 table'}
            >
                <thead>
                <tr className={'table-header'}>
                    <th className={'h3'}>Unique Ref. No.</th>
                    <th className={'h3'}>Date</th>
                    <th className={'h3'}>Description</th>
                    <th className={'h3'}>Location</th>
                    <th className={'h3'}>Priority</th>
                    <th className={'h3'}>Status</th>
                </tr>
                </thead>
                <tbody>
                {/* The data is received back as an array of objects. So the Array.map method is used to map loop
                 through the array and display the data of the individual job objects */}
                {jobs.map((job, index) => {
                    /* Only non-archived documents will be displayed */
                    if (job.archive === false) {
                        return (
                            <tr key={index} className={'table-row'}>
                                <td className={'h5'}>{job._id}</td>
                                <td className={'h5'}>{(job.date).slice(0, 10)}</td>
                                <td className={'h5'}>{job.description}</td>
                                <td className={'h5'}>{job.location}</td>
                                <td className={'h5'}>{job.priority}</td>
                                <td className={'h5'}>{job.status}</td>
                            </tr>
                        )
                    } else {
                        return (
                            <tr>
                                <td colSpan={7}>
                                </td>
                            </tr>
                        )
                    }
                })}
                </tbody>
            </Table>
        </div>
    )
}