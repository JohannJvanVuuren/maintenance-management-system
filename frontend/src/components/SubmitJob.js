/* Import of the React hooks needed in this component */
import {useState} from "react";
/* Import of the axios client to handle API requests to the backend */
import axios from 'axios';

/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';

/* Definition of the SubmitJob component that will be used to initially complete details
* about and submit a job for tracking */
export const SubmitJob = () => {

    /* Declaration and initialisation of state variables needed */
    const [date, setDate] = useState('');
    const [description, setDescription] = useState('');
    const [location, setLocation] = useState('');
    const [priority, setPriority] = useState('');
    const [status, setStatus] = useState('');


    /* The change handler for the radio buttons for priority of the request */
    const radioButtonHandler = (event) => {
        setPriority(event.target.value);
    }

    const selectHandler = (event) => {
        setStatus(event.target.value);
    }

    /* The form's submit handler which also changes the status to 'Submitted' */
    const submitHandler = (event) => {

        setStatus('Submitted');
        console.log('click')
        /* Configurations for the axios request to the backend that will send the information to be
        * stored in the database */
        const url = 'http://localhost:8000/saveNewJob';
        const config = {
            date: date,
            description: description,
            location: location,
            priority: priority,
            status: status
        }

        /* The axios request to the backend */
        axios.post(url, config)
            .then(response => {
                return response.data
            })
            .catch(error => {
                console.log('Error', error.message);
            })
    }

    /* The rendering of the submit job input form */
    return (
        <div className={'submit-job-container'}>
            <h1 className={'section-titles'}>Submit Task</h1>
            <form
                className={'submit-form'}
                onSubmit={submitHandler}
            >
                <div className={'submit-form-textfields'}>
                    <label htmlFor={'date'}>
                        Date:
                        <input
                            type={'date'}
                            id={'date'}
                            name={'date'}
                            onChange={(event) => setDate(new Date(event.target.value))}
                            required
                        />
                    </label>
                    <label htmlFor={'description'}>
                        Description:
                        <input
                            type={'text'}
                            id={'description'}
                            name={'description'}
                            placeholder={'Description'}
                            onChange={(event) => setDescription(event.target.value)}
                            style={{
                                fontWeight: 900,
                                color: '#800000'
                            }}
                            required
                        />
                    </label>
                    <label htmlFor={'location'}>
                        Location:
                        <input
                            type={'text'}
                            id={'location'}
                            name={'location'}
                            placeholder={'Location'}
                            onChange={(event) => setLocation(event.target.value)}
                            style={{
                                fontWeight: 900,
                                color: '#800000'
                            }}
                            required
                        />
                    </label>
                </div>
                {/*<p>Priority</p>*/}

                <div className={'submit-form-radio-buttons'}>
                    <p>Priority</p>
                    <label
                        htmlFor={'urgent'}
                        className={'label-radio'}
                    >
                        Urgent
                        <input
                            type={'radio'}
                            id={'urgent'}
                            name={'priority'}
                            value={'Urgent'}
                            onChange={radioButtonHandler}
                            // checked={priority === 'urgent'}
                        />
                    </label>
                    <label
                        htmlFor={'routine'}
                        className={'label-radio'}
                        /* Inline style to override React-Bootstrap interference */
                        style={{marginLeft: 1 + 'rem'}}
                    >
                        Routine
                        <input
                            className={'radio-second'}
                            type={'radio'}
                            id={'routine'}
                            name={'priority'}
                            value={'Routine'}
                            onChange={radioButtonHandler}
                            // checked={priority === 'routine'}
                        />
                    </label>
                </div>
                <div className={'submit-form-select'}>
                    <label
                        htmlFor={'status'}
                        className={'select-label'}
                        /* Inline styling to override React-Bootstrap default styles */
                        style={{
                            marginTop: 1 + 'rem',
                            marginLeft: 4 + 'rem'
                        }}
                    >
                        Status:
                        <select
                            id={'status'}
                            name={'status'}
                            onChange={selectHandler}
                            style={{
                                fontWeight: 900,
                                color: '#800000'
                            }}
                        >
                            <option>--Select--</option>
                            <option value={'Submitted'}>Submitted</option>
                            <option value={'In progress'}>In progress</option>
                            <option value={'Completed'}>Completed</option>
                        </select>
                    </label></div>
                <input
                    className={'btn-submit'}
                    type={'submit'}
                    value={'Submit'}
                />
            </form>
        </div>
    )
}