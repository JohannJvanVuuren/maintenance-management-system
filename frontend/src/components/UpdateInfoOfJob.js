/* Import of the axios client to handle requests to the backend */
import axios from 'axios';
/* Import of Link component from React Router to transfer state information to UpdateJobForm */
import {Link} from 'react-router-dom';
/* Import of React hooks used in this component */
import {useEffect, useState} from "react";

/* Import of main stylesheet generated by SCSS and preprocessor */
import '../scss/main.css';
/* Import of React-Bootstrap styled components */
import Table from 'react-bootstrap/Table';
import {dateFormatter} from "../helperFunctions/dateFormatter";

/* Definition of the ListJobsByDate component. The purpose of this component is to simply display
* all jobs in the database */
export const UpdateInfoOfJob = () => {

    /* Declaration and initialisation of the state variable needed in this component */
    const [jobs, setJobs] = useState([]);

    /* A useEffect hook to run the axios call to the backend to retrieve the required information from
    * the database */
    useEffect(() => {

        /* Setup of configuration of the axios call */
        const url = 'http://localhost:8000/getJobListByDate';
        const config = {
            field: 'date'
        }

        /* axios call to the .../getJobListByDate endpoint to get the information that needs to be displayed
        * in this component */
        axios.get(url, config)
            .then(response => {
                setJobs(response.data);
            })
            .catch(error => {
                console.log('Error', error.message);
            })
    }, []); //No dependencies, so the hook will only run once when the page is rendered

    return (
        <div>
            <h1 className={'section-titles'}>Update Details</h1>
            {/* Rendering of a React-bootstrap table suited for this purpose */}
            <Table
                striped bordered hover variant="dark"
                responsive
                className={'mt-5 table'}
            >
                <thead>
                <tr className={'table-header'}>
                    <th className={'h3'}>Unique Ref. No.</th>
                    <th className={'h3'}>Date</th>
                    <th className={'h3'}>Description</th>
                    <th className={'h3'}>Location</th>
                    <th className={'h3'}>Priority</th>
                    <th className={'h3'}>Status</th>
                    <th className={'h3'}>Click to update</th>
                </tr>
                </thead>
                <tbody>
                {/* The data is received back as an array of objects. So the Array.map method is used to map loop
                 through the array and display the data of the individual job objects */}
                {jobs.map((job, index) => {
                    /* Only non-archived documents are displayed */
                    if (job.archive === false) {
                        return (
                            <tr key={index} className={'table-row'}>
                                <td className={'h5'}>{job._id}</td>
                                <td className={'h5'}>{dateFormatter(job.date)}</td>
                                <td className={'h5'}>{job.description}</td>
                                <td className={'h5'}>{job.location}</td>
                                <td className={'h5'}>{job.priority}</td>
                                <td className={'h5'}>{job.status}</td>
                                <td>
                                    <Link
                                        to={'/updateJobForm'}
                                        className={'btn-update'}
                                        state={{
                                            id: job._id,
                                            date: job.date,
                                            description: job.description,
                                            location: job.location,
                                            priority: job.priority,
                                            status: job.status
                                        }}
                                    >
                                        <button
                                            className={'btn-update'}
                                        >
                                            Update
                                        </button>
                                    </Link>
                                </td>
                            </tr>
                        )
                    } else {
                        return (
                            <tr>
                                <td colSpan={7}>
                                </td>
                            </tr>
                        )
                    }
                })}
                </tbody>
            </Table>
        </div>
    )
}