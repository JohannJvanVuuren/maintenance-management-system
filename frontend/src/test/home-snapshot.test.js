/* Import of the renderer module from react-test-renderer */
import renderer from 'react-test-renderer';

/* Import of component to be tested */
import { LandingPage} from "../components/LandingPage";

/* Snapshot test of the LandingPage component */
describe('Snapshot test of the LandingPage component', () => {
    it('LandingPage component renders correctly', () => {
        const tree = renderer.create(<LandingPage/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})