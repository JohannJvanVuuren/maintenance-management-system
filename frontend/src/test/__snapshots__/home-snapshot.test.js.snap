// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of the LandingPage component LandingPage component renders correctly 1`] = `
<div
  className="landing-container"
>
  <h1>
    Welcome
  </h1>
  <p>
    This system assists with the management of maintenance jobs and their tracking.
  </p>
  <p>
    Navigation through the system takes place though the navigation buttons at the top of the screen.
  </p>
  <p>
    To return to the home page click on the title at the top of the page.
  </p>
</div>
`;
